quic/s2n-quic-platform/src/socket/io/tx.rs

```diff
@@ -0,0 +1,439 @@
+        // take the first free descriptor, we should have at least one item
+    }
+}
+
+/// Tracks the current state of a GSO message
+#[derive(Debug, Default)]
+pub struct GsoSegment<Handle> {
+    handle: Handle,
+    ecn: ExplicitCongestionNotification,
+    count: usize,
+    size: usize,
+}
+
+pub struct TxQueue<'a, T: Message> {
+    channels: &'a mut [Producer<T>],
+    channel_index: usize,
+    message_index: usize,
+    pending_release: u32,
+    gso_segment: Option<GsoSegment<T::Handle>>,
+    max_segments: usize,
+    max_mtu: usize,
+    capacity: usize,
+    is_full: &'a mut bool,
+}
+
+impl<'a, T: Message> TxQueue<'a, T> {
+    /// Tries to send a message as a GSO segment
+    ///
+    /// Returns the Err(Message) if it was not able to. Otherwise, the index of the GSO'd message is returned.
+    #[inline]
+    fn try_gso<M: tx::Message<Handle = T::Handle>>(
+        &mut self,
+        mut message: M,
+    ) -> Result<Result<tx::Outcome, M>, tx::Error> {
+        // the message doesn't support GSO to return it
+        if !T::SUPPORTS_GSO {
+            return Ok(Err(message));
+        }
+
+        let max_segments = self.max_segments;
+
+        let (prev_message, gso) = if let Some(gso) = self.gso_message() {
+            gso
+        } else {
+            return Ok(Err(message));
+        };
+
+        debug_assert!(
+            max_segments > 1,
+            "gso_segment should only be set when max_gso > 1"
+        );
+
+        // check to make sure the message can be GSO'd and can be included in the same
+        // GSO payload as the previous message
+        let can_gso = message.can_gso(gso.size, gso.count)
+            && message.path_handle().strict_eq(&gso.handle)
+            && message.ecn() == gso.ecn;
+
+        // if we can't use GSO then flush the current message
+        if !can_gso {
+            self.flush_gso();
+            return Ok(Err(message));
+        }
+
+        debug_assert!(
+            gso.count < max_segments,
+            "{} cannot exceed {}",
+            gso.count,
+            max_segments
+        );
+
+        let payload_len = prev_message.payload_len();
+
+        unsafe {
+            // Safety: all payloads should have enough capacity to extend max_segments *
+            // gso.size
+            prev_message.set_payload_len(payload_len + gso.size);
+        }
+
+        // allow the message to write up to `gso.size` bytes
+        let buffer = &mut T::payload_mut(prev_message)[payload_len..];
+        let buffer = tx::PayloadBuffer::new(buffer);
+
+        match message.write_payload(buffer, gso.count).and_then(|size| {
+            // we don't want to send empty packets
+            if size == 0 {
+                Err(tx::Error::EmptyPayload)
+            } else {
+                Ok(size)
+            }
+        }) {
+            Err(err) => {
+                unsafe {
+                    // revert the len to what it was before
```

Comment:
hmm good question... we definitely can but I don't think it'll really get used since if we panic the whole endpoint is closed, at least in its current implementation.

Reply:
I refactored the code to not modify the message, which makes it so we don't have to do any kind of reverts.